### PR TITLE
task-8: integration with RDS

### PR DESCRIPTION
**Pull request by nslcorp**
Book shop: https://d1pcilzlaus1qs.cloudfront.net/
Base backend url: https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/
Front End MR link: https://github.com/nslcorp/book-shop-client/pull/7

### Ping status requests
GET: https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/
GET: https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/ping

### "Carts" requests
.. NOTE: all requests require Basic authorisation
need to add header: `Authorization` and value: `Basic dXNlcjp1c2Vy` what means `user:user` in base64 format.

**GET:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/cart

**PUT:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/cart

**DELETE:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/cart

### "Order" requests
.. NOTE: all requests require Basic authorisation
need to add header: `Authorization` and value: `Basic dXNlcjp1c2Vy` what means `user:user` in base64 format.

**GET:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/order

**PUT:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/order

```
body example:
{
    "total": "125",
    "items":[
        {"productId":"c9f4c1b6-573f-4551-82a5-eaaaa5f37a4b","count":2}
    ],
    "address":{"comment":"some comment","address":"22233","lastName":"111","firstName":"222"}
}
```

**DELETE:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/cart/:existed-order-id

example
**DELETE:** https://xy0xbhhiid.execute-api.eu-central-1.amazonaws.com/dev/api/profile/cart/66c3fcfc-c825-44b6-b86c-ddcef54553b9

## Progress status:

DONE:  Task 8.1 ==> copy and run Fork a Next template repository.
DONE:  Task 8.2 ==> create RDS instance with PostgreSQL, create tables in DB
DONE:  Task 8.3 ==> update source code in the application to use PostgreSQL instead of memory storage.
DONE:  Task 8.4 ==> commit all code to Serverless, check that everything works

DONE:  Client application should display alerts for the responses in 401 and 403 HTTP statuses. 

DONE:  +4 (All languages) - Create users table and integrate with it
DONE:  +3 (All languages) - Integrate Cart service with FE repository
DONE: +20 (All languages) - Create orders table and integrated with it Order model:

NOT-DONE: +3 (All languages) - Transaction based creation of checkout

P.S Fixed issue with add items to cart. Now count calculates correctly
P.S add validation for Order. now backend responses 4xx with for different cases that could course.

### CODE REVIEW

File: cart-service/sql/ddl-queries.sql [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+);
+
+-- Carts
+create type status_enum as enum ('OPEN', 'ORDERED')
+
+DROP TABLE if exists carts
+create table if not exists carts (
+   id uuid primary key default uuid_generate_v4(),
+   user_id uuid default uuid_generate_v4(),
+   created_at date not null,
```
Comment (relnofollow) [inline cart-service/sql/ddl-queries.sql:17]: Please add 'updated_at' column to the table, as it's mentioned in the task description

image

File: cart-service/sql/ddl-queries.sql [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   foreign key ("cart_id") references "carts" ("id"),
+   product_id uuid,
+   count integer not null default 1
+);
+
+
+DROP TABLE if exists products;
+create table if not exists products (
+   id uuid,
+   foreign key ("id") references "cart_items" ("product_id"),
```
Comment (relnofollow) [inline cart-service/sql/ddl-queries.sql]: I'm not sure, but maybe this foreign key constraint should exist in the 'cart_items' table, not in the 'products' table? In other words, products can exist without cart items, while cart items can't exist without products, and different cart items may reference same product.

File: cart-service/src/cart/services/cart.service.ts [hunk 3]
```diff
@@ -34,22 +46,58 @@ export class CartService {
...
-  updateByUserId(userId: string, { items }: Cart): Cart {
-    const { id, ...rest } = this.findOrCreateByUserId(userId);
-
-    const updatedCart = {
-      id,
-      ...rest,
-      items: [ ...items ],
+  async updateByUserId(userId: string, item: LegacyCartItem): Promise<any> {
+    const { id, ...rest } = await this.findOrCreateByUserId(userId);
+    const cart = await this.findOrCreateByUserId(userId);
```
Comment (relnofollow) [inline cart-service/src/cart/services/cart.service.ts]: There're two consecutive calls of `findOrCreateByUserId`: in this line and in the line above

File: cart-service/src/cart/services/cart.service.ts [hunk 4]
```diff
@@ -34,22 +46,58 @@ export class CartService {
...
+    console.log('cart', cart);
+
+    let cartItem = cart.items.find(
+      (record) => record.productId === item.product.id,
+    );
+    if (cartItem) {
+      if (item.count === 0) {
+        await this.cartItemRepository.remove(cartItem);
```
Comment (relnofollow) [inline cart-service/src/cart/services/cart.service.ts:58]: Probably, a 'return' statement is missing here

File: cart-service/serverless.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+service: cart-service
+
+plugins:
+  - serverless-offline
+  - serverless-dotenv-plugin
```
Comment (relnofollow) [inline cart-service/serverless.yml:5]: According to the serverless documentation, there's no need to use this plugin anymore:
https://www.serverless.com/plugins/serverless-dotenv-plugin#do-you-need-this-plugin

File: cart-service/src/cart/services/cart.service.ts [hunk 6]
```diff
@@ -34,22 +46,58 @@ export class CartService {
...
 
-  removeByUserId(userId): void {
-    this.userCarts[ userId ] = null;
-  }
+  async removeByUserId(userId): Promise<void> {
+    const cart = await this.cartRepository.findOne({ userId });
+    const cartItem = await this.cartItemRepository.find({ cartId: cart.id });
+    const productIDs = cartItem.map((record) => record.productId);
 
+    const pResponse = await this.productRepository.delete({
```
Comment (relnofollow) [inline cart-service/src/cart/services/cart.service.ts]: I'm not sure that product should be deleted, when a cart item is deleted. 
For example, 'user1' wants to buy book 'Book1' and 'user2' wants to buy 'Book1'. When a cart for 'user1' is deleted, then the cart of 'user2' seems to be in invalid state.

File: cart-service/src/order/order.controller.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    private cartService: CartService,
+  ) {}
+
+  @UseGuards(BasicAuthGuard)
+  @Put()
+  async create(@Req() req: AppRequest, @Body() body) {
+    const userId = getUserIdFromRequest(req);
+    const { address, items } = body as OrderPayload;
+
+    if (!address || !items) {
```
Comment (relnofollow) [inline cart-service/src/order/order.controller.ts]: I'm not sure, but it seems that 'items' property is not used while creating an order

**Comment by relnofollow** [comment]
Could you please guide me, what is the expected body for 'PUT' request for 'order'?
I've tried such JSON with Authorization header provided for user/user pair, but it gives me 500 error:

```
{
    "product": {
        "id": "9011866e-3aa1-443e-97c1-914c3d3220b1",
        "title": "The Culture Map [Erin Meyer]",
        "description": "An international business expert helps you understand and navigate cultural differences in this insightful and practical guide, perfect for both your work and personal life.",
        "price": 13.99
    },
    "count": 1
}
```

cr-8

File: cart-service/serverless.yml (hunk 5 above)
Comment (nslcorp) [reply cart-service/serverless.yml:5]: As you can see I do not duplicate env names in serverless.yaml file
plugin is do the following: "This plugin is only useful if you want to automatically import all variables from .env into functions:"

File: cart-service/src/cart/services/cart.service.ts (hunk 3 above)
Comment (nslcorp) [reply cart-service/src/cart/services/cart.service.ts]: Yes. Absolutely right. Dummy mistake :(

File: cart-service/src/cart/services/cart.service.ts (hunk 4 above)
Comment (nslcorp) [reply cart-service/src/cart/services/cart.service.ts:58]: Correct. Thanks

File: cart-service/src/cart/services/cart.service.ts (hunk 6 above)
Comment (nslcorp) [reply cart-service/src/cart/services/cart.service.ts]: Yes. It looks you are righ.
To say more this point was not so clear to me by business rules, generally we should not keep product here, we have special service to keep product(s) and get them my ID.

but by logic that you described and as User1Cart and User2Cart actually link to the same product.

I will "remove" deleting products.

File: cart-service/src/order/order.controller.ts (hunk 7 above)
Comment (nslcorp) [reply cart-service/src/order/order.controller.ts]: Yes. It looks you are right. and in DB we don't use it. I will remove it from validation and from TS interfase.

File: cart-service/sql/ddl-queries.sql (hunk 1 above)
Comment (nslcorp) [reply cart-service/sql/ddl-queries.sql:17]: Done.

**Comment by nslcorp** [comment]
> 



> Could you please guide me, what is the expected body for 'PUT' request for 'order'? I've tried such JSON with Authorization header provided for user/user pair, but it gives me 500 error:
> 
> ```
> {
>     "product": {
>         "id": "9011866e-3aa1-443e-97c1-914c3d3220b1",
>         "title": "The Culture Map [Erin Meyer]",
>         "description": "An international business expert helps you understand and navigate cultural differences in this insightful and practical guide, perfect for both your work and personal life.",
>         "price": 13.99
>     },
>     "count": 1
> }
> ```
> 
> cr-8 [ cr-8 ](https://user-images.githubusercontent.com/4612163/231527231-a34705b1-5a7e-4adf-8d73-48715616855b.gif) [ ](https://user-images.githubusercontent.com/4612163/231527231-a34705b1-5a7e-4adf-8d73-48715616855b.gif)



Here is request example:

```{
    "total": "125",
    "items":[
        {"productId":"c9f4c1b6-573f-4551-82a5-eaaaa5f37a4b","count":2}
    ],
    "address":{"comment":"some comment","address":"22233","lastName":"111","firstName":"222"}
}
```

**Pull request merged**